narwhal/types/src/tests/primary_type_tests.rs

```diff
@@ -0,0 +1,100 @@
+use rand::{rngs::StdRng, SeedableRng};
+use signature::Signer;
+
+use crate::{BatchDigest, CertificateDigest, Header, HeaderDigest, Metadata};
+
+fn arb_keypair() -> impl Strategy<Value = KeyPair> {
+    (any::<[u8; 32]>())
+        .prop_map(|rand| {
+            let mut rng = StdRng::from_seed(rand);
+            KeyPair::generate(&mut rng)
+        })
+        .no_shrink()
+}
+
+fn clean_signed_header(kp: KeyPair) -> impl Strategy<Value = Header> {
+    (
+        any::<u64>(),
+        any::<u64>(),
+        collection::vec((BatchDigest::arbitrary(), any::<u32>()), 0..10),
+        collection::vec(CertificateDigest::arbitrary(), 0..10),
+    )
+        .prop_map(move |(round, epoch, batches, parents)| {
+            let payload = batches
+                .into_iter()
+                .map(|(batch_digest, worker_id)| (batch_digest, worker_id as WorkerId))
+                .collect();
+
+            let parents = parents.into_iter().collect();
+
+            let header = Header {
+                author: kp.public().clone(),
+                round,
+                epoch,
+                payload,
+                parents,
+                id: OnceCell::default(),
+                signature: Signature::default(),
+                metadata: Metadata::default(),
+            };
+            Header {
+                id: OnceCell::with_value(header.digest()),
+                signature: kp.sign(Digest::from(header.digest()).as_ref()),
+                ..header
+            }
+        })
+}
+
+fn arb_signed_header(kp: KeyPair) -> impl Strategy<Value = Header> {
+    (
+        clean_signed_header(kp.copy()),
+        HeaderDigest::arbitrary(),
+        any::<usize>(),
+    )
+        .prop_map(move |(clean_header, random_digest, naughtiness)| {
+            let naughtiness = naughtiness % 100;
+            if naughtiness < 95 {
+                clean_header
+            } else if naughtiness < 99 {
+                let signature = kp.sign(Digest::from(random_digest).as_ref());
+                // naughty: we provide a well-signed random header
+                Header {
+                    id: OnceCell::with_value(random_digest),
+                    signature,
+                    ..clean_header
+                }
+            } else {
+                // naughty: we provide an ill-signed random header
+                Header {
+                    id: OnceCell::with_value(random_digest),
+                    ..clean_header
+                }
+            }
+        })
+}
+
+fn arb_header() -> impl Strategy<Value = Header> {
+    arb_keypair().prop_flat_map(arb_signed_header)
+}
+
+proptest! {
+    #[test]
+    fn header_deserializes_to_correct_id(header in arb_header()) {
+        let serialized = bincode::serialize(&header).unwrap();
+        let deserialized: Header = bincode::deserialize(&serialized).unwrap();
+        // we may not have header.digest() == header.id(), due to the naughty cases above
```

Comment:
What does that mean? Should our test fail in this case? Actually, do we expect it to fail?

Reply:
The naughty headers are specially crafted so that their `id` is populated with a wrong id. They are malformed, since a correct header always has `foo.digest() == foo.id()`.

We check here that deserializing a header, even a malformed one, produces a correctly-formed header in all cases.